mteb/evaluation/evaluators/InstructionRetrievalEvaluator.py

```diff
@@ -0,0 +1,263 @@
+import tqdm
+
+import pytrec_eval
+from sentence_transformers import SentenceTransformer
+from sentence_transformers.models import Transformer, WordEmbeddings
+import torch
+from collections import defaultdict
+
+from .Evaluator import Evaluator
+from .utils import cos_sim, dot_score, mrr, recall_cap, hole, top_k_accuracy
+
+logger = logging.getLogger(__name__)
+
+# Adapted from https://github.com/beir-cellar/beir/blob/f062f038c4bfd19a8ca942a9910b1e0d218759d4/beir/retrieval/search/dense/exact_search.py#L12
+class DenseRetrievalExactSearch:    
+    def __init__(self, model, batch_size: int = 128, corpus_chunk_size: int = 50000, **kwargs):
+        # Model is class that provides encode_corpus() and encode_queries()
+        self.model = model
+        self.batch_size = batch_size
+        self.score_functions = {'cos_sim': cos_sim, 'dot': dot_score}
+        self.score_function_desc = {'cos_sim': "Cosine Similarity", 'dot': "Dot Product"}
+        self.corpus_chunk_size = corpus_chunk_size
+        self.show_progress_bar = kwargs.get("show_progress_bar", True)
+        self.convert_to_tensor = kwargs.get("convert_to_tensor", True)
+        self.results = {}
+        self.save_corpus_embeddings = kwargs.get("save_corpus_embeddings", False)
+        self.corpus_embeddings = defaultdict(list)
+
+    def search(self, 
+               corpus: Dict[str, Dict[str, str]], 
+               queries: Dict[str, str], 
+               instructions: Dict[str, str],
+               top_k: int, 
+               score_function: str,
+               return_sorted: bool = False, 
+               **kwargs) -> Dict[str, Dict[str, float]]:
+        
+        # reranking here, so no need for the heap
+        top_k = len(corpus)
+        qid = list(queries.keys())[0] 
+        assert len(queries) == 1
+        
+        # Create embeddings for all queries using model.encode_queries()
+        # Runs semantic search against the corpus embeddings
+        # Returns a ranked list with the corpus ids
+        if score_function not in self.score_functions:
+            raise ValueError("score function: {} must be either (cos_sim) for cosine similarity or (dot) for dot product".format(score_function))
+            
+        logger.info("Encoding Queries...")
+        query_ids = list(queries.keys())
+        self.results = {qid: {} for qid in query_ids}
+        queries = [queries[qid] for qid in queries]
+        query_embeddings = self.model.encode_queries(
+            queries, batch_size=self.batch_size, show_progress_bar=self.show_progress_bar, convert_to_tensor=self.convert_to_tensor, instructions=instructions)
+          
+        logger.info("Sorting Corpus by document length (Longest first)...")
+
+        corpus_ids = sorted(corpus, key=lambda k: len(corpus[k].get("title", "") + corpus[k].get("text", "")), reverse=True)
+        corpus = [corpus[cid] for cid in corpus_ids]
+
+        logger.info("Encoding Corpus in batches... Warning: This might take a while!")
```

Comment:
```suggestion
        logger.info("Encoding Corpus in batches. This might take a while!")
```
Warning seems to suggest that something might be wrong.